[REVIEW] null supported iterator

kovaltan opened the pull request:
This PR provide custom input iterator which supports column with nulls.

Some of current cudf reduction and aggregation functions have to write a kernel to support nulls since Thrust and cub does not treat nulls which consists of null bitmaps, while Thrust and cub are iterator driven APIs.

The aim of this PR is to provice iterator which is easy to use for reduction, aggregation implementation to replace the custom kernels into simple Thrust/cub call.

The input iterator is implemented using `thrust::iterator_adaptor` and `thrust::counting_iterator`.
When creating a null supported input iterator, the iterator requires pointer of data and null bitmap and an `identity` value like below.

```
auto it_dev = cudf::make_iterator<true>(static_cast<T*>( column->data ), 
    column->valid, T{0});
```

`*(it_dev +id)` returns `static_cast<T*>( column->data )[id]` if `column->valid` at `id` is true, and T{0} if `column->valid` at `id` = false.
The `identity` value depends on the data type and the type of aggregation.
```
e.g.
T = int32_t and aggregation is a kind of `sum`:
   identity = T{0}
T = cudf::wapper and aggregation is a kind of `max`:
   identity = std::numeric_limits<T>::lowest()
```

Example of use case:
```
template <typename T>
T sum_of_column(const gdf_column& col)
{
  // have to separate the call since the returned iterator type is different 
  // if col.valid == nullptr or not
  if( col.valid == nullptr){
    auto it_dev = cudf::make_iterator<false, T>(col, T{0});
    return thrust::reduce(thrust::device, it_dev, it_dev + col.size, T{0}, 
        cudf::DeviceSum{});
  }else{
    auto it_dev = cudf::make_iterator<true, T>(col, T{0});
    return thrust::reduce(thrust::device, it_dev, it_dev + col.size, T{0},
        cudf::DeviceSum{});
  }
}
```

The column input iterator itself returns only a scalar value of y he data at id or identity value.
thrust::make_transform_iterator can trasform the iterator output into various forms and use cases, like up casting, squared value, struct of values, etc...

Examples of use cases:
1. template parameter for same precision input iterator
`     auto it = make_iterator<has_nulls, T>(column, T{0});`

2. template parameter for upcasting input iterator
```
    auto it = make_iterator<has_nulls, T, T_upcast>(column, T{0});
```

3. template parameter for upcasting + squared input iterator
``` 
    cudf::transformer_squared<T_upcast> transformer{};
    auto it = make_iterator<has_nulls, T, T_upcast>(column, T{0});
    auto it_squared = thrust::make_transform_iterator(it, transformer);
```

4. template parameter for using `meanvar`
```
    using T_output = cudf::meanvar<T_upcast>;
    cudf::transformer_meanvar<T_upcast> transformer{};
    auto it_pair = make_pair_iterator<has_nulls, T, T_upcast>(column, T{0});
    auto it_meanvar = thrust::make_transform_iterator(it_pair, transformer);
```

5. template parameter for custom indexed iterator
```
    gdf_index_type *indices;
    auto it = make_iterator<has_nulls, T, T, T_index*>(column, T{0}, indices);
```


Task list:
- [x] iterator test with/without nulls
- [x] test with sum of square iterator
- [x] iterator with index array test
- [x] up cast test case (reduction `char` input => `double` output)
- [x] struct case test case (sum to `ColumnDataOutputs`)
- [x] add utility functions `cudf::iterator::make_XXX`
- [x] refactoring
- [x] documentation
- [x] performance check


References:
    Thrust iterators: https://thrust.github.io/doc/group__fancyiterator.html
    CUB reduction:  https://nvlabs.github.io/cub/structcub_1_1_device_reduce.html
    Thrust reduction: https://thrust.github.io/doc/group__reductions.html

jrhemstad commented on cpp/tests/iterator/iterator_test.cu:
```diff
@@ -0,0 +1,393 @@
+// --------------------------------------------------------------------------------------------------------
+
+
+void gen_nullbitmap(std::vector<gdf_valid_type>& v, std::vector<bool>& host_bools)
```
I think this function can be removed if you just use `column_wrapper` in your tests.

kovaltan replied:
sure.

jrhemstad commented on cpp/tests/iterator/iterator_test.cu:
```diff
@@ -0,0 +1,393 @@
+    std::vector<bool> host_bools({1, 1, 0, 1, 1, 1, 0, 1, 1});
+    std::vector<gdf_valid_type> host_nulls;
+    gen_nullbitmap(host_nulls, host_bools);
+    thrust::device_vector<gdf_valid_type> dev_nulls(host_nulls);
```
I think this would simpler if you used `column_wrapper`.

harrism replied:
Here's an example: https://github.com/rapidsai/cudf/blob/branch-0.8/cpp/tests/stream_compaction/stream_compaction_tests.cu

kovaltan replied:
sure.

kovaltan replied:
@jrhemstad @harrism 
Is it possible to get `column_wrapper::bitmask` to create `host_nulls`?
I also test reduction with `thrust::host `, `host_nulls` is required for testing host memory iterator.

kovaltan replied:
self resolved.
`column_wrapper::to_host()` would work.

jrhemstad commented on cpp/tests/iterator/iterator_test.cu:
```diff
@@ -0,0 +1,393 @@
+
+
+// column_input_iterator
+template<typename T_output, typename T_input, typename Iterator=thrust::counting_iterator<gdf_index_type> >
```
So making sure I understand this...

`T_output` is passed into `column_input_iterator` so that you can specify the type of `super_t::reference`?

kovaltan replied:
Yes, the 6th (= last -1) template parameter of  the parent `thrust::iterator_adaptor` is the `super_t::reference`.

https://github.com/rapidsai/cudf/blob/c971adf6dbf87e337a9d0db8ed535259f38b3e78/cpp/tests/iterator/iterator_test.cu#L108-L111

jrhemstad commented on cpp/tests/iterator/iterator_test.cu:
```diff
@@ -0,0 +1,393 @@
+};
+
+template<typename T>
+struct ColumnData{
```
In a production version of this design, instead of having both `ColumnDataNonNull` and `ColumnData`, I'd add a `bool` template argument to `ColumnData` indicating whether or not the `valid` mask is present. In that way, you can avoid the performance hit from always checking `gdf_is_valid` by checking `if(nulls_present)` if there aren't any nulls in the column. Since it's a template argument, and known at compile time, the compiler can optimize out the `if` check if `nulls_present == false`.

kovaltan replied:
Those 2 structs are different in the count of the members.
I'm not sure if I set bool template argument `nulls_present`, the compiler optimized out the unused struct member if `nulls_present` = false?
If the member is not optimized, the copy of iterator may take unnecessary copies.

jrhemstad replied:
You could explicitly specialize for `true` and `false` and then you can have different members of the class depending on the template arg.

kovaltan replied:
That sounds good/

jrhemstad replied:
That said, I don't know if it's necessary. Having an unused 8B pointer member of the iterator probably isn't going to have much overhead, is it?

kovaltan replied:
This is addressed in latest commit. Please check.

jrhemstad commented on cpp/tests/iterator/iterator_test.cu:
```diff
@@ -0,0 +1,393 @@
+    For group_by.cumsum() (scan base group_by) may not be single pass scan.
+    There is a possiblity that this process may be used for group_by.cumsum().
+*/
+TYPED_TEST(IteratorTest, group_by_iterator)
```
I don't understand how this test relates to `group_by`. This just looks to be doing a reduction, same as the other tests.

kovaltan replied:
This does a reduction for the selected indices.
Let's say, .group_by('A') where 'A' = [1, 1, 2,  1, 2, 1,  2, 2, 2]
The indices for `1` => [0, 1, 3, 5], `2` => [2, 4, 6, 7, 8]
We actually do 2 reductions for the 2 indices instead of distributed single reduction.

I know group_by.sum() never uses reduction for the selected indices.
However, how do we implement group_by.cumsum() (scan base group_by)?
I feel there is a possibility that this process may be used for group_by.cumsum().
Of course, this may not be used if there is better way.

jrhemstad replied:
Oh! I see, I missed what was going on here.

You're passing an array of `[0,1,3,5]` as the base iterator for `column_input_iterator` which will only perform the reduction on those indices! Makes sense now. 

That said, I think this would end up being an expensive way to implement a groupby since you'd be doing a separate kernel call for each unique group. Furthermore, you'd have to first construct a list of indices where values belonging to each group reside. 

When I've thought about implementing a `groupby.cumsum()`, I'd thought to first sort the data, and then use a `thrust::exclusive_scan_by_key`.

kovaltan replied:
> That said, I think this would end up being an expensive way to implement a groupby since you'd be doing a separate kernel call for each unique group. Furthermore, you'd have to first construct a list of indices where values belonging to each group reside.
> 
> When I've thought about implementing a `groupby.cumsum()`, I'd thought to first sort the data, and then use a `thrust::exclusive_scan_by_key`.

`thrust::exclusive_scan_by_key` looks good. I missed it.
Okay, I will leave this test exist since it test the feasibility of indexed access. But I will add a comment "this is a feasibility test, this will be slow and may not be used in cudf real case."

jrhemstad commented on cpp/tests/iterator/iterator_test.cu:
```diff
@@ -0,0 +1,393 @@
+    friend class thrust::iterator_core_access;
+
+    __host__ __device__
+    column_input_iterator &operator=(const column_input_iterator &other)
```
Is this necessary?

kovaltan replied:
I',m still investigating if it is necessary.
Actually, I've referenced the implementation of `thrust::transform_iterator`.
https://github.com/thrust/thrust/blob/cc4be34c4b3bf7a4b5a7ef2977bab5bfcc676878/thrust/iterator/transform_iterator.h#L245-L297

harrism commented on cpp/tests/iterator/iterator_test.cu:
```diff
@@ -0,0 +1,393 @@
+template<typename T>
+struct ColumnData{
+    const T *data;
+    const gdf_valid_type *valid;
```
Can we use bit_mask_t here instead? And bit_mask::is_valid() below instead of gdf_is_valid? We should be phasing out gdf_is_valid IMO in favor of the 32-bit version for better performance.

kovaltan commented:
@jrhemstad 

> Making sure I understand. The basic idea here is to create an iterator over the elements of a column. If the column element is valid, it returns that element, otherwise, returns an `identity` value.

Exactly.

> In this way, all of the code in `reductions.cu` and `scan.cu` could be eliminated in favor of just a single `thrust/cub` reduce/scan call? Sounds pretty great!

Yes, I believe the current `recuction.cu` and `scan.cu` could be replaced by a single cub call.

harrism commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,312 @@
+template <typename T_element, typename T_output = T_element,
+    typename T_output_helper = ColumnOutputSingle<T_output>,
+    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
+auto make_iterator_with_nulls(const T_element *_data, const bit_mask::bit_mask_t *_valid,
```
What about factories that take columns? 
```
auto make_iterator(gdf_column const& column) {...
```

kovaltan replied:
The iterator class with and without nulls must be separated at compile time.
Taking `gdf_column` is possible, but we need a pair of `make_iterator_with_nulls` and `make_iterator_without_nulls` for `gdf_column` input.
If it is okay, I will add a pair of factories from `gdf_column`.

kovaltan replied:
I've added a pair of `make_iterator_with_nulls` and `make_iterator_without_nulls` for `gdf_column `input.
Please check it.
Tomorrow I will add test cases for `make_iterator_XXX` with gdf_column input.

harrism replied:
Soon(ish) we will have a `cudf::column` class. Presumably these iterators will just be members. Will that not work if it's not known at compile time whether the column has nulls?

kovaltan replied:
@harrism

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,443 @@
+
+/** -------------------------------------------------------------------------*
+ * @brief helper struct to output a  squared scalar value
+ * A helper output struct for `column_input_iterator`
```
I think this struct needs some clearer documentation on what it is used for.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,443 @@
+
+// @overload ColumnInput<T_output, T_element, true>
+template<typename T_output, typename T_element>
+struct ColumnInput<T_output, T_element, true>{
```
I don't think `ColumnInput` should have different member functions depending on `nulls_present`. That makes it difficult to use this struct uniformly whether or not it has nulls. `nulls_present` should simply be an optimization that avoids the `is_valid` check.

kovaltan replied:
Agree, will update.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,443 @@
+template <typename T_element, typename T_output = T_element,
+    typename T_output_helper = ColumnOutputSingle<T_output>,
+    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
+auto make_iterator_with_nulls(const T_element *data, const bit_mask::bit_mask_t *valid,
```
Instead of having `make_iterator_with_nulls` and `make_iterator_no_nulls`, just add a `bool has_nulls` template argument. 

So instead of `make_iterator_with_nulls`, it's just `make_iterator<true>` or `make_iterator<false>` instead of `make_iterator_without_nulls`.

kovaltan replied:
@jrhemstad 
`just add a bool has_nulls template argument` sounds a good idea, but I have a concern.
It is possible to add `bool has_nulls` as 1st template argument, and can create a specialized case make_iterator<true or false>.
However, since all of `make_iterator_xxx` actually  have different parameters.
So the specialization case for each parameter set is only one of <true or false>.
If we don't mind that, I will do it. Do you mind?

Another option (though I don't like this)
Indeed if there is `identity` parameter, it creates null iterator, else, it creates non-null iterator.
We can distinguish null or non-null iterator by `identity` parameter.
But I feel it is ambiguous, I think we should explicitly set null or non-null at creating the iterator.

jrhemstad replied:
> It is possible to add bool has_nulls as 1st template argument, and can create a specialized case make_iterator.

Since `make_iterator...` has additional template parameters, you're talking about partially specialization, which isn't possible with a free function like this.

I see no need to specialize on the `has_nulls` template parameter. Furthermore, I don't think there is an issue with having unused constructor parameters if `has_nulls==false`. Just default the unused parameters and I think it'll be fine. 

```
template <bool has_nulls, typename T_element, typename T_output = T_element,
    typename T_output_helper = ColumnOutputSingle<T_output>,
    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
auto make_iterator(const T_element *data, const bit_mask::bit_mask_t *valid = nullptr,
    T_element identity = T_element{}, Iterator_Index const it = Iterator_Index(0))
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,504 @@
+
+    CUDA_HOST_DEVICE_CALLABLE
+    T_output at(gdf_index_type id) const {
+        return T_output(get_value(id), is_valid(id));
```
This is computing `is_valid` twice here. How about just get rid of the `get_value` and `is_valid` private members and do this: 

```
    T_output at(gdf_index_type id) const {
        return is_valid(valid, id) ? T_output(data[id], true) : T_output(identity, false); 
    }
```

kovaltan replied:
Fixed.

jrhemstad replied:
It looks like this hasn't actually been changed yet.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,504 @@
+ * If nulls_present = false, the valid flag corresponds to null bitmask flag
+ * at `id`, and the data value = (is_valid(id))? data[id] : identity;
+ *
+ * @tparam  T_output a scalar data type to be output
```
This needs a clearer name than `T_output`. Furthermore, it is not a scalar data type. It's some type that can be constructed with arguments `(T, bool)`. 

Likewise, it needs clearer documentation on what this object does. In reality, it's more of a "Loader" or "Mutator" object, that given some value, potentially mutates it into a new value. 

Finally, and most importantly, does `ColumnInput` really need to be complicated with this mutator object? From a design perspective, I believe `ColumnInput` should be simplified such that it does only one thing: for a requested value at index `i`, if the value is valid, return it. Else, return an `identity` value. 

Any more complex functionality using your mutator objects should be built _on top of_ this simplified `ColumnInput`. Right now, they are intertwined and I think it makes the implementation harder to follow.

kovaltan commented:
@jrhemstad @harrism 
Please re-review this, I've updated from your review.
Thanks!

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,459 @@
+ * @tparam  T_input a scalar data type to be input
+ * -------------------------------------------------------------------------**/
+template<typename T>
+struct mutator_single
```
Likewise, you use `mutator_single` as the default in `make_iterator`. 

I think you should add an `identity_mutator` that does not modify the source element whatsoever and use that as the default.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,459 @@
+    }
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
```
Why does this `mutator` need an `operator==`, but none of the rest? Make this clear when you document the required interface of every `mutator` class.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,459 @@
+ * @tparam  T_input a scalar data type to be input
+ * -------------------------------------------------------------------------**/
+template<typename T>
+struct mutator_single
```
What still bothers me about the `mutator` objects is that there is a expected interface that all `mutator` objects should implement, but currently there's no description nor enforcement of that interface.  

For example, every `mutator` must have a constructor template with a signature like `(T, bool)`, or an implicit conversion operator to `T`. 

Obviously we don't want to use an abstract class with pure-virtual functions to define an interface. We could maybe use CRTP to define and enforce an interface. But at the very least, the `mutator` objects warrant a thorough comment block describing what they are for and their expected interface.

kovaltan commented:
@jrhemstad 

I've removed `mutator` and introduced `transformer`. While `mutator` holds value and data conversion logic, 
`transformer` only has the transform logic in operator(). It simply transforms `thrust::pair<T, bool>` into `T_output` form.

https://github.com/rapidsai/cudf/blob/8d38b8f83d77f7933d98e197881f21beb6798df1/cpp/src/iterator/iterator.cuh#L138-L146

For `transformer_meanvar`, I speraterd the `transformer` and the output struct.
I believe it is cleaner than before.
The struct 'meanvar' has necessary operators and constructor, that's the reason old 'mutator_meanvar' had different operators than others.

https://github.com/rapidsai/cudf/blob/8d38b8f83d77f7933d98e197881f21beb6798df1/cpp/src/iterator/iterator.cuh#L50-L53
https://github.com/rapidsai/cudf/blob/8d38b8f83d77f7933d98e197881f21beb6798df1/cpp/src/iterator/iterator.cuh#L183-L194

kovaltan commented:
I did a perf test.
Roughly raw pointer and iterator is almost same performance (within the margin of error ).
"raw stream cub" computes cub::DeviceReduce::Reduce by raw pointer.
"iterator cub <false>" computes cub::DeviceReduce::Reduce by cudf::make_iterator<false>().
"iterator cub <true>" computes cub::DeviceReduce::Reduce by cudf::make_iterator<true>().

**$ ./ITERATOR_BENCH 42000000 10000 d**
Iterator performance test:
  column_size = 42000000
  num iterates = 10000

GDF_FLOAT64
new allocation: 0
raw strem cub: 		0.000583246
iterator cub <false>: 	0.000582897
iterator cub <true>: 	0.000591545
new allocation: 1
raw strem cub: 		0.00058339
iterator cub <false>: 	0.000583103
iterator cub <true>: 	0.00059166

kovaltan commented:
I've added thrust and cudf::reduction(at branch-0.7 using cub::BlockReduce + atomics) cases at benchmark.
The result shows:

1. "raw pointer" and "iterator" redution are same perf.

  raw strem cub: 		0.000583391
  iterator cub <false>: 	0.000582988

  raw strem thust: 		0.000780092
  iterator thust <false>: 	0.000778671

2. "iterator with nulls" may be ~1.5% slower than "iterator without nulls"

  It is expected since there is additional operation if the iterator has nulls
  0.000591559/0.000582784 = 1.015 = 1.5%

3. cub::DeviceReduce with iterator may be 3x faster than current cudf::reduction

  0.00171958/0.000582784 = 2.95 = 3x slower

Below is the raw number of an result.


$ ./ITERATOR_BENCH 42000000 10000 d
Iterator performance test:
  column_size = 42000000
  num iterates = 10000

GDF_FLOAT64
new allocation: 0
raw strem cub: 		0.000583509
iterator cub <false>: 	0.000582784
iterator cub <true>: 	0.000591559
new allocation: 1
raw strem cub: 		0.000583391
iterator cub <false>: 	0.000582988
iterator cub <true>: 	0.000591546
raw strem thust: 		0.000780092
iterator thust <false>: 	0.000778671
iterator thust <true>: 	0.000786775
raw stream cub::BlockReduce<false>: 	0.00171958
raw stream cub::BlockReduce<true>: 	0.00173041

kovaltan commented:
@jrhemstad @harrism 
Could you re-review again?
The benchmark shows "raw pointer" and "iterator" redution are same perf, the iterator driven reduction is apparently faster than current implementation.

I also filed a performance issue #1953 there is significant performance issue at GDF_INT8/16 and this PR will resolve it.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,389 @@
+};
+
+// ---------------------------------------------------------------------------
+// transformers: transforms thrust::pair<T_element, bool> into output form.
```
How about something like:
```/** -------------------------------------------------------------------------*
 * @brief Construct an instance of `T_output` using a `thrust::pair<T_element, bool>
 * 
 * Uses a scalar, boolean pair to construct a new object.
 * -------------------------------------------------------------------------**/
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,389 @@
+ * @tparam  T_output  a scalar data type of output
+ * -------------------------------------------------------------------------**/
+template<typename T_element, typename T_output=T_element>
+struct transformer
```
I'd call this the `scalar_cast_transformer` as it just casts the scalar from one type to another.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,389 @@
+// transformers: transforms thrust::pair<T_element, bool> into output form.
+
+/** -------------------------------------------------------------------------*
+ * @brief transformer to output a scalar value
```
```suggestion
 * @brief Transforms a scalar by casting it to another scalar type
 * 
 * By default, performs an identity cast, i.e., casts to the scalar's original type.
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,389 @@
+};
+
+/** -------------------------------------------------------------------------*
+ * @brief transformer to output a squared scalar value
```
```suggestion
 * @brief Transforms a scalar by first casting to another type, and then squaring the result. 
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,389 @@
+};
+
+/** -------------------------------------------------------------------------*
+ * @brief transformer to output a set of values
```
```suggestion
 * @brief Uses a scalar value to construct a `meanvar` object. 
```

harrism commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,395 @@
+ * If has_nulls = false, the valid flag corresponds to null bitmask flag
+ * at `id`, and the data value = (is_valid(id))? data[id] : identity;
+ *
+ * @tparam  T_element a native element type (cudf data type) of input element array
```
```suggestion
 * @tparam  T_element cudf data type of input element array
```

harrism commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,395 @@
+// helper functions to make iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief helper function to make iterator with nulls
```
```suggestion
 * @brief helper function to make a cudf column iterator
```

harrism commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,395 @@
+}
+
+/** -------------------------------------------------------------------------*
+ *  @overload auto make_iterator(const T_element *data,
```
I think overload documentation should say what this overload does (e.g. how it is different).

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,406 @@
+    // shorthand for the name of the iterator_adaptor we're deriving from
+    using super_t = thrust::iterator_adaptor<
+      column_input_iterator<T_iterator_output, T_column_input, Iterator>, Iterator,
+      T_iterator_output, thrust::use_default, thrust::use_default, T_iterator_output, thrust::use_default
```
Like Michal said, I think you should leave the 6th template argument as the default. Since it will default to `T_iterator_output&`.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+ * @tparam  T_element cudf data type of input element array
+ * @tparam  T_output  cudf data type of output and `identity` value
+ *                    which is used when null bitmaps flag is false.
+ * @tparam  has_nulls if true, this struct holds only data array.
```
This doc is reversed. If `has_nulls` is true, then it had data and bitmask.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+
+/** -------------------------------------------------------------------------*
+ * @brief pair accessor with/without null bitmask
+ * A unary function returns pair of scalar value and validity at `id`.
```
```suggestion
 * A unary function returns `thrust::pair<ResultT, bool>`. If the element at index `i` is valid, returns `ResultT{data[i]}` and `true` indicating the value was valid. If the element at `i` is null, returns `ResultT{identity}` and `false` indicating the element was null.
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+  bit_mask::bit_mask_t const* bitmask{};
+  T_output const identity{};
+
+  value_accessor(T_element const* e, bit_mask::bit_mask_t const* b, T_output i)
```
Document constructor, especially describe the `identity` parameter.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+struct value_accessor<T_element, T_output, false>
+{
+  T_element const* elements{};
+  value_accessor(T_element const* e, bit_mask::bit_mask_t const*, T_output) : elements{e} {}
```
Document. Describe that the `bitmask` and `identity` parameters are unused.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+  }
+};
+
+template <typename T_element, typename T_output>
```
Document as specialization for columns without nulls.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+struct value_accessor;
+
+template <typename T_element, typename T_output>
+struct value_accessor<T_element, T_output, true>
```
Document as specialization for columns that contain null values.

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+ * Input iterator which can be used for cub and thrust.
+ * The iterator returns same cudf data type of input: `T_element`.
+ *
+ * @tparam has_nulls True if the data has valid bit mask, False else
```
```suggestion
 * @tparam has_nulls Indicates if the column contains null values (`null_count > 0`)
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+ * The iterator returns same cudf data type of input: `T_element`.
+ *
+ * @tparam has_nulls True if the data has valid bit mask, False else
+ * @tparam T_element The cudf data type of input array
```
```suggestion
 * @tparam ElementType The element's data type
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+ *
+ * @tparam has_nulls True if the data has valid bit mask, False else
+ * @tparam T_element The cudf data type of input array
+ * @tparam T_output  cudf data type of output and `identity` value
```
```suggestion
 * @tparam ResultType  The data type returned by de-referencing the iterator (e.g., `*it`) 
```

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+ * -------------------------------------------------------------------------**/
+template <bool has_nulls, typename T_element, typename T_output = T_element,
+    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
+auto make_iterator(const T_element *data, const bit_mask::bit_mask_t *valid,
```
I'd default `valid` and `identity`, because it doesn't make sense to have to provide arguments for those if `has_nulls==false`.  Then add a `if(has_nulls == true)` to ensure that arguments were passed in.

kovaltan replied:
@jrhemstad 
>  Then add a if(has_nulls == true) to ensure that arguments were passed in.
How can ensure that `identity` is passed in when if(has_nulls == true)?
For bitmask, the constructor of ` value_accessor<ElementType, ResultType, true>` check it.

kovaltan replied:
I've added  default parameter for `valid` and `identity`, then validation check for `data`, `valid` pointer. There is no check for `identity`. See below:
https://github.com/rapidsai/cudf/blob/6d5e3ffe0029ef002a35808f69aa1108331cc5db/cpp/src/iterator/iterator.cuh#L262-L273

jrhemstad commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,322 @@
+// helper functions to make iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief helper function to make a cudf column iterator
```
```suggestion
 * @brief Constructs an iterator over the elements of a column.
 * 
 * If the column contains no null values (indicated by `has_nulls == false`) then dereferencing an iterator `it` returned by this function as `*(it + n)` will return `ResultType{ static_cast<ElementType*>(data)[n] }`.
 
 * If the column contains null values (indicated by `has_nulls == true`) then the result of de-referencing an iterator `it` returned by this function as `*(it+n)` will depend if element is valid or null. If the element is valid, it will return `ResultType{ static_cast<ElementType*>(data)[n] }`. If the element is null, it will return `ResultType{identity}`.
```

harrism commented:
@kovaltan To help finalize can you click "resolve" on all review comments that you have addressed? Thanks!

harrism commented on cpp/src/iterator/iterator.cuh:
```diff
@@ -0,0 +1,382 @@
+#endif
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE
```
Should this be host-callable? I believe the bitmask and the elements are always device pointers, aren't they?

kovaltan replied:
At first, I add thrust host test case for debugging and make sure if the iterator works well. It's already confirmed, then __device__ should be enough.
I've changed the value_accessor and pair_accessor intp `CUDA_DEVICE_CALLABLE`.

But for functor in transform_unary_functions.cuh, I still use `CUDA_HOST_DEVICE_CALLABLE` since there may be possibility to use them at host code.

kovaltan commented:
rerun test

kovaltan commented:
rerun tests
